nexus/test-utils/src/lib.rs

```diff
@@ -139,247 +141,626 @@ pub async fn test_setup<N: NexusServer>(
+        let dns = dns_server::InMemoryServer::new(&log).await.unwrap();
+
+        let SocketAddr::V6(address) = dns.dropshot_server.local_addr() else {
+            panic!("Unsupported IPv4 DNS address");
+        };
+        self.rack_init_builder.add_service(
+            address,
+            ServiceKind::InternalDns,
+            internal_dns::ServiceName::InternalDns,
+            sled_id,
+        );
+
+        self.internal_dns = Some(dns);
+    }
+
+    pub fn build(self) -> ControlPlaneTestContext<N> {
+        ControlPlaneTestContext {
+            start_time: self.start_time,
+            server: self.server.unwrap(),
+            external_client: self.external_client.unwrap(),
+            internal_client: self.internal_client.unwrap(),
+            database: self.database.unwrap(),
+            clickhouse: self.clickhouse.unwrap(),
+            sled_agent_storage: self.sled_agent_storage.unwrap(),
+            sled_agent: self.sled_agent.unwrap(),
+            oximeter: self.oximeter.unwrap(),
+            producer: self.producer.unwrap(),
+            logctx: self.logctx,
+            dendrite: self.dendrite.unwrap(),
+            external_dns_zone_name: self.external_dns_zone_name.unwrap(),
+            external_dns: self.external_dns.unwrap(),
+            internal_dns: self.internal_dns.unwrap(),
+            silo_name: self.silo_name.unwrap(),
+            user_name: self.user_name.unwrap(),
+        }
+    }
+
+    pub async fn teardown(self) {
+        if let Some(server) = self.server {
+            server.close().await;
+        }
+        if let Some(mut database) = self.database {
+            database.cleanup().await.unwrap();
+        }
+        if let Some(mut clickhouse) = self.clickhouse {
+            clickhouse.cleanup().await.unwrap();
+        }
+        if let Some(sled_agent) = self.sled_agent {
+            sled_agent.http_server.close().await.unwrap();
+        }
+        if let Some(oximeter) = self.oximeter {
+            oximeter.close().await.unwrap();
+        }
+        if let Some(producer) = self.producer {
+            producer.close().await.unwrap();
+        }
+        if let Some(mut dendrite) = self.dendrite {
+            dendrite.cleanup().await.unwrap();
+        }
+        self.logctx.cleanup_successful();
     }
 }
 
+pub async fn test_setup_with_config<N: NexusServer>(
+    test_name: &str,
+    config: &mut omicron_common::nexus_config::Config,
+    sim_mode: sim::SimMode,
+    initial_cert: Option<Certificate>,
+) -> ControlPlaneTestContext<N> {
+    let mut builder =
+        ControlPlaneTestContextBuilder::<N>::new(test_name, config);
+
+    builder.start_crdb().await;
+    builder.start_clickhouse().await;
+    builder.start_dendrite().await;
+    builder.start_internal_dns().await;
+    builder.start_external_dns().await;
+    builder.start_nexus_internal().await;
+    builder.start_sled(sim_mode).await;
+    builder.start_crucible_pantry().await;
```

Comment:
nice for readability to wrap these steps up, even though this is pretty much doing what it was doing before